include/grpcpp/security/tls_credentials_options.h

```diff
@@ -0,0 +1,298 @@
+};
+
+/** TLS credential reload arguments, wraps grpc_tls_credential_reload_arg. It is
+ * used for experimental purposes for now and it is subject to change.
+ *
+ * The credential reload arg contains all the info necessary to schedule/cancel
+ * a credential reload request. The callback function must be called after
+ * finishing the schedule operation. See the description of the
+ * grpc_tls_credential_reload_arg struct in grpc_security.h for more details.
+ * **/
+class TlsCredentialReloadArg {
+ public:
+  /** TlsCredentialReloadArg does not take ownership of the C arg that is passed
+   * to the constructor. One must remember to free any memory allocated to the C
+   * arg after using the setter functions below. **/
+  TlsCredentialReloadArg(grpc_tls_credential_reload_arg* arg) : c_arg_(arg) {}
+  ~TlsCredentialReloadArg();
+
+  /** Getters for member fields. The callback function is not exposed.
+   * They return the corresponding fields of the underlying C arg. In the case
+   * of the key materials config, it creates a new instance of the C++ key
+   * materials config from the underlying C grpc_tls_key_materials_config. **/
+  void* cb_user_data() const;
+  std::shared_ptr<TlsKeyMaterialsConfig> key_materials_config() const;
+  grpc_ssl_certificate_config_reload_status status() const;
+  grpc::string error_details() const;
+
+  /** Setters for member fields. They modify the fields of the underlying C arg.
+   * The setters for the key_materials_config and the error_details allocate
+   * memory when modifying c_arg_, so one must remember to free c_arg_'s
+   * original key_materials_config or error_details after using the appropriate
+   * setter function.
+   * **/
+  void set_cb_user_data(void* cb_user_data);
+  void set_key_materials_config(
+      const std::shared_ptr<TlsKeyMaterialsConfig>& key_materials_config);
+  void set_status(grpc_ssl_certificate_config_reload_status status);
+  void set_error_details(const grpc::string& error_details);
+
+  /** Calls the C arg's callback function. **/
+  void OnCredentialReloadDoneCallback();
+
+ private:
+  grpc_tls_credential_reload_arg* c_arg_;
+};
+
+/** An interface that the application derives and uses to instantiate a
+ * TlsCredentialReloadConfig instance. Refer to the definition of the
+ * grpc_tls_credential_reload_config in grpc_tls_credentials_options.h for more
+ * details on the expectations of the member functions of the interface. **/
+struct TlsCredentialReloadInterface {
+  virtual ~TlsCredentialReloadInterface() = default;
+  /** A callback that invokes the credential reload. **/
+  virtual int Schedule(TlsCredentialReloadArg* arg) { return 1; }
+  /** A callback that cancels a credential reload request. **/
+  virtual void Cancel(TlsCredentialReloadArg* arg) {}
+  /** A callback that cleans up any data associated to the
+   * interface or the config. It will be called when the config is no longer
+   * using the interface. **/
+  virtual void Release() {}
```

Comment:
I've left the `Release` method, since the user can use it to clean up memory that they may have allocated during e.g. the `Schedule` method. As you pointed out, this is basically the only reason it is needed now, as the config owns the interface now.

Reply:
Since `Release` is called in the Config's dtor, I would guess whatever is needed in `Release` can be just moved to the dtor of the interface. Thus there is no reason to keep the method on the interface. Do you have a concrete example of having something that has to be in `Release`?

Reply:
Great point - I had not thought to leave this to the destructor of the interface. I will remove the `Release` methods in the next commit (after we sort out the other comment).